File: Treachery-iOS/Treachery-iOSTests/GameBoardViewModelTests.swift
import Testing
import Foundation
@testable import Treachery_iOS

@MainActor
struct GameBoardViewModelTests {

    // MARK: - Helpers

    private func makeSampleGame(
        state: GameState = .inProgress,
        gameMode: GameMode = .treachery,
        hostId: String = "host1",
        winningTeam: String? = nil
    ) -> Game {
        Game(
            id: "game1", code: "ABCD", hostId: hostId, state: state,
            gameMode: gameMode, maxPlayers: 8, startingLife: 40,
            winningTeam: winningTeam, playerIds: ["host1", "u1", "u2", "u3"],
            createdAt: Date()
        )
    }

    private func makeSamplePlayers() -> [Player] {
        [
            Player(id: "p0", orderId: 0, userId: "host1", displayName: "Host",
                   role: .leader, identityCardId: "card1", lifeTotal: 40,
                   isEliminated: false, isUnveiled: false, joinedAt: Date()),
            Player(id: "p1", orderId: 1, userId: "u1", displayName: "Alice",
                   role: .assassin, identityCardId: "card2", lifeTotal: 40,
                   isEliminated: false, isUnveiled: false, joinedAt: Date()),
            Player(id: "p2", orderId: 2, userId: "u2", displayName: "Bob",
                   role: .traitor, identityCardId: "card3", lifeTotal: 40,
                   isEliminated: true, isUnveiled: true, joinedAt: Date()),
            Player(id: "p3", orderId: 3, userId: "u3", displayName: "Carol",
                   role: .assassin, identityCardId: nil, lifeTotal: 40,
                   isEliminated: false, isUnveiled: false, joinedAt: Date()),
        ]
    }

    private func makeVM(
        game: Game? = nil,
        players: [Player] = [],
        currentUserId: String? = "u1"
    ) -> GameBoardViewModel {
        let g = game ?? makeSampleGame()
        let p = players.isEmpty ? makeSamplePlayers() : players
        return GameBoardViewModel(
            gameId: g.id,
            previewPlayers: p,
            previewGame: g,
            currentUserId: currentUserId,
            firestoreManager: MockFirestoreManager(),
            cloudFunctions: MockCloudFunctions(),
            cardDatabase: CardDatabase(cards: []),
            planeDatabase: PlaneDatabase(cards: [])
        )
    }

    // MARK: - Current Player

    @Test func currentPlayerMatchesUserId() {
        let vm = makeVM(currentUserId: "u1")
        #expect(vm.currentPlayer?.userId == "u1")
        #expect(vm.currentPlayer?.displayName == "Alice")
    }

    @Test func currentPlayerNilWhenNoMatch() {
        let vm = makeVM(currentUserId: "nobody")
        #expect(vm.currentPlayer == nil)
    }

    @Test func currentPlayerNilWhenNoUserId() {
        let vm = makeVM(currentUserId: nil)
        #expect(vm.currentPlayer == nil)
    }

    // MARK: - Game State

    @Test func isGameFinished() {
        let vm = makeVM(game: makeSampleGame(state: .finished))
        #expect(vm.isGameFinished == true)
    }

    @Test func isGameNotFinished() {
        let vm = makeVM(game: makeSampleGame(state: .inProgress))
        #expect(vm.isGameFinished == false)
    }

    @Test func winningTeamParsesRole() {
        let vm = makeVM(game: makeSampleGame(winningTeam: "assassin"))
        #expect(vm.winningTeam == .assassin)
    }

    @Test func winningTeamNilWhenNoWinner() {
        let vm = makeVM(game: makeSampleGame(winningTeam: nil))
        #expect(vm.winningTeam == nil)
    }

    // MARK: - Alive Players

    @Test func alivePlayersExcludesEliminated() {
        let vm = makeVM()
        let alive = vm.alivePlayers
        #expect(alive.count == 3) // Bob is eliminated
        #expect(!alive.contains(where: { $0.displayName == "Bob" }))
    }

    // MARK: - Game Mode Booleans

    @Test func treacheryModeFlags() {
        let vm = makeVM(game: makeSampleGame(gameMode: .treachery))
        #expect(vm.isTreacheryActive == true)
        #expect(vm.isPlanechaseActive == false)
    }

    @Test func planechaseModeFlags() {
        let vm = makeVM(game: makeSampleGame(gameMode: .planechase))
        #expect(vm.isTreacheryActive == false)
        #expect(vm.isPlanechaseActive == true)
    }

    @Test func bothModeFlags() {
        let vm = makeVM(game: makeSampleGame(gameMode: .treacheryPlanechase))
        #expect(vm.isTreacheryActive == true)
        #expect(vm.isPlanechaseActive == true)
    }

    // MARK: - Host Detection

    @Test func isHostWhenMatches() {
        let vm = makeVM(game: makeSampleGame(hostId: "u1"), currentUserId: "u1")
        #expect(vm.isHost == true)
    }

    @Test func isNotHostWhenDifferent() {
        let vm = makeVM(game: makeSampleGame(hostId: "host1"), currentUserId: "u1")
        #expect(vm.isHost == false)
    }

    // MARK: - Role Visibility

    @Test func canSeeOwnRole() {
        let vm = makeVM(currentUserId: "u1")
        let alice = vm.players.first { $0.userId == "u1" }!
        #expect(vm.canSeeRole(of: alice) == true)
    }

    @Test func canSeeUnveiledPlayerRole() {
        let vm = makeVM(currentUserId: "u1")
        let bob = vm.players.first { $0.displayName == "Bob" }!
        #expect(bob.isUnveiled)
        #expect(vm.canSeeRole(of: bob) == true)
    }

    @Test func canSeeLeaderRole() {
        let vm = makeVM(currentUserId: "u1")
        let host = vm.players.first { $0.role == .leader }!
        #expect(vm.canSeeRole(of: host) == true)
    }

    @Test func cannotSeeHiddenNonLeaderRole() {
        let vm = makeVM(currentUserId: "u1")
        let carol = vm.players.first { $0.displayName == "Carol" }!
        #expect(!carol.isUnveiled)
        #expect(carol.role != .leader)
        #expect(carol.userId != "u1")
        #expect(vm.canSeeRole(of: carol) == false)
    }

    // MARK: - Die Roll Cost

    @Test func dieRollCostFirstRollFree() {
        var game = makeSampleGame(gameMode: .treacheryPlanechase)
        game.planechase = PlanechaseState(
            useOwnDeck: false, currentPlaneId: "p1", usedPlaneIds: [],
            lastDieRollerId: nil, dieRollCount: 0
        )
        let vm = makeVM(game: game)
        #expect(vm.dieRollCost == 0)
    }

    @Test func dieRollCostSecondRollFree() {
        var game = makeSampleGame(gameMode: .treacheryPlanechase)
        game.planechase = PlanechaseState(
            useOwnDeck: false, currentPlaneId: "p1", usedPlaneIds: [],
            lastDieRollerId: nil, dieRollCount: 1
        )
        let vm = makeVM(game: game)
        #expect(vm.dieRollCost == 0)
    }

    @Test func dieRollCostThirdRollCostsOne() {
        var game = makeSampleGame(gameMode: .treacheryPlanechase)
        game.planechase = PlanechaseState(
            useOwnDeck: false, currentPlaneId: "p1", usedPlaneIds: [],
            lastDieRollerId: nil, dieRollCount: 2
        )
        let vm = makeVM(game: game)
        #expect(vm.dieRollCost == 1)
    }
}

File: TreacheryAndroid/app/src/test/java/com/solomon/treachery/viewmodel/GameBoardViewModelTests.kt
package com.solomon.treachery.viewmodel

import androidx.lifecycle.SavedStateHandle
import com.google.firebase.Timestamp
import com.solomon.treachery.mocks.*
import com.solomon.treachery.model.*
import com.solomon.treachery.ui.game.GameBoardViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.*
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

@OptIn(ExperimentalCoroutinesApi::class)
class GameBoardViewModelTests {

    private val testDispatcher = UnconfinedTestDispatcher()
    private lateinit var firestore: MockFirestoreRepository
    private lateinit var cloudFunctions: MockCloudFunctionsRepository
    private lateinit var cardDb: MockCardDatabase
    private lateinit var planeDb: MockPlaneDatabase

    private val now = Timestamp.now()

    @BeforeEach
    fun setUp() {
        Dispatchers.setMain(testDispatcher)
        firestore = MockFirestoreRepository()
        cloudFunctions = MockCloudFunctionsRepository()
        cardDb = MockCardDatabase()
        planeDb = MockPlaneDatabase()
    }

    @AfterEach
    fun tearDown() {
        Dispatchers.resetMain()
    }

    private fun makeVM(gameId: String = "game-1"): GameBoardViewModel {
        val savedState = SavedStateHandle(mapOf("gameId" to gameId))
        return GameBoardViewModel(firestore, cloudFunctions, cardDb, planeDb, savedState).also {
            it.currentUserId = "user-1"
        }
    }

    private fun makeGame(
        state: GameState = GameState.IN_PROGRESS,
        gameMode: GameMode = GameMode.TREACHERY,
        hostId: String = "user-1",
        winningTeam: String? = null,
        planechase: PlanechaseState? = null,
    ) = Game(
        id = "game-1", code = "ABCD", hostId = hostId,
        state = state, gameMode = gameMode,
        maxPlayers = 8, startingLife = 40, playerIds = listOf("user-1", "user-2"),
        createdAt = now, winningTeam = winningTeam, planechase = planechase,
    )

    private fun makePlayers() = listOf(
        Player(id = "p-1", orderId = 0, userId = "user-1", displayName = "Host",
            role = Role.LEADER, identityCardId = "card-1", lifeTotal = 40,
            isEliminated = false, isUnveiled = false, joinedAt = now),
        Player(id = "p-2", orderId = 1, userId = "user-2", displayName = "Player 2",
            role = Role.ASSASSIN, identityCardId = "card-2", lifeTotal = 35,
            isEliminated = false, isUnveiled = false, joinedAt = now),
        Player(id = "p-3", orderId = 2, userId = "user-3", displayName = "Player 3",
            role = Role.TRAITOR, identityCardId = "card-3", lifeTotal = 0,
            isEliminated = true, isUnveiled = true, joinedAt = now),
        Player(id = "p-4", orderId = 3, userId = "user-4", displayName = "Player 4",
            role = Role.GUARDIAN, identityCardId = "card-4", lifeTotal = 40,
            isEliminated = false, isUnveiled = true, joinedAt = now),
    )

    // ── currentPlayer ──

    @Nested
    inner class CurrentPlayerTests {

        @Test
        fun `returns correct player for userId`() = runTest {
            val vm = makeVM()
            firestore.playersFlowSource.value = makePlayers()
            advanceUntilIdle()

            assertEquals("user-1", vm.currentPlayer?.userId)
            assertEquals("Host", vm.currentPlayer?.displayName)
        }

        @Test
        fun `returns null when userId not in players`() = runTest {
            val vm = makeVM()
            vm.currentUserId = "not-in-game"
            firestore.playersFlowSource.value = makePlayers()
            advanceUntilIdle()

            assertNull(vm.currentPlayer)
        }
    }

    // ── Game state ──

    @Nested
    inner class GameStateTests {

        @Test
        fun `isGameFinished true when finished`() = runTest {
            val vm = makeVM()
            firestore.gameFlowSource.value = makeGame(state = GameState.FINISHED)
            advanceUntilIdle()

            assertTrue(vm.isGameFinished)
        }

        @Test
        fun `isGameFinished false when in_progress`() = runTest {
            val vm = makeVM()
            firestore.gameFlowSource.value = makeGame(state = GameState.IN_PROGRESS)
            advanceUntilIdle()

            assertFalse(vm.isGameFinished)
        }

        @Test
        fun `winningTeam parses role`() = runTest {
            val vm = makeVM()
            firestore.gameFlowSource.value = makeGame(winningTeam = "assassin")
            advanceUntilIdle()

            assertEquals(Role.ASSASSIN, vm.winningTeam)
        }

        @Test
        fun `winningTeam null when no winner`() = runTest {
            val vm = makeVM()
            firestore.gameFlowSource.value = makeGame()
            advanceUntilIdle()

            assertNull(vm.winningTeam)
        }
    }

    // ── Alive players ──

    @Nested
    inner class AlivePlayersTests {

        @Test
        fun `excludes eliminated players`() = runTest {
            val vm = makeVM()
            firestore.playersFlowSource.value = makePlayers()
            advanceUntilIdle()

            assertEquals(3, vm.alivePlayers.size)
            assertTrue(vm.alivePlayers.none { it.isEliminated })
        }

        @Test
        fun `returns all when none eliminated`() = runTest {
            val vm = makeVM()
            firestore.playersFlowSource.value = makePlayers().map { it.copy(isEliminated = false) }
            advanceUntilIdle()

            assertEquals(4, vm.alivePlayers.size)
        }
    }

    // ── Game mode flags ──

    @Nested
    inner class GameModeFlagTests {

        @Test
        fun `treachery mode flags`() = runTest {
            val vm = makeVM()
            firestore.gameFlowSource.value = makeGame(gameMode = GameMode.TREACHERY)
            advanceUntilIdle()

            assertTrue(vm.isTreacheryActive)
            assertFalse(vm.isPlanechaseActive)
        }

        @Test
        fun `planechase mode flags`() = runTest {
            val vm = makeVM()
            firestore.gameFlowSource.value = makeGame(gameMode = GameMode.PLANECHASE)
            advanceUntilIdle()

            assertFalse(vm.isTreacheryActive)
            assertTrue(vm.isPlanechaseActive)
        }

        @Test
        fun `combined mode flags`() = runTest {
            val vm = makeVM()
            firestore.gameFlowSource.value = makeGame(gameMode = GameMode.TREACHERY_PLANECHASE)
            advanceUntilIdle()

            assertTrue(vm.isTreacheryActive)
            assertTrue(vm.isPlanechaseActive)
        }

        @Test
        fun `none mode flags`() = runTest {
            val vm = makeVM()
            firestore.gameFlowSource.value = makeGame(gameMode = GameMode.NONE)
            advanceUntilIdle()

            assertFalse(vm.isTreacheryActive)
            assertFalse(vm.isPlanechaseActive)
        }
    }

    // ── Host detection ──

    @Nested
    inner class HostDetectionTests {

        @Test
        fun `isHost true when userId matches`() = runTest {
            val vm = makeVM()
            vm.currentUserId = "user-1"
            firestore.gameFlowSource.value = makeGame(hostId = "user-1")
            advanceUntilIdle()

            assertTrue(vm.isHost)
        }

        @Test
        fun `isHost false when userId does not match`() = runTest {
            val vm = makeVM()
            vm.currentUserId = "user-2"
            firestore.gameFlowSource.value = makeGame(hostId = "user-1")
            advanceUntilIdle()

            assertFalse(vm.isHost)
        }
    }

    // ── Role visibility ──

    @Nested
    inner class RoleVisibilityTests {

        @Test
        fun `can always see own role`() = runTest {
            val vm = makeVM()
            vm.currentUserId = "user-1"
            firestore.playersFlowSource.value = makePlayers()
            advanceUntilIdle()

            val ownPlayer = vm.players.value.find { it.userId == "user-1" }!!
            assertTrue(vm.canSeeRole(ownPlayer))
        }

        @Test
        fun `can see unveiled player role`() = runTest {
            val vm = makeVM()
            vm.currentUserId = "user-1"
            firestore.playersFlowSource.value = makePlayers()
            advanceUntilIdle()

            val unveiledPlayer = vm.players.value.find { it.userId == "user-4" }!!
            assertTrue(unveiledPlayer.isUnveiled)
            assertTrue(vm.canSeeRole(unveiledPlayer))
        }

        @Test
        fun `can see leader role even if not unveiled`() = runTest {
            val vm = makeVM()
            vm.currentUserId = "user-2"
            firestore.playersFlowSource.value = makePlayers()
            advanceUntilIdle()

            val leaderPlayer = vm.players.value.find { it.role == Role.LEADER }!!
            assertFalse(leaderPlayer.isUnveiled)
            assertTrue(vm.canSeeRole(leaderPlayer))
        }

        @Test
        fun `cannot see hidden non-leader role`() = runTest {
            val vm = makeVM()
            vm.currentUserId = "user-1"
            firestore.playersFlowSource.value = makePlayers()
            advanceUntilIdle()

            val hiddenAssassin = vm.players.value.find { it.userId == "user-2" }!!
            assertFalse(hiddenAssassin.isUnveiled)
            assertNotEquals(Role.LEADER, hiddenAssassin.role)
            assertFalse(vm.canSeeRole(hiddenAssassin))
        }
    }

    // ── Planechase computed properties ──

    @Nested
    inner class PlanechaseTests {

        @Test
        fun `dieRollCost is 0 when no rolls`() = runTest {
            val vm = makeVM()
            firestore.gameFlowSource.value = makeGame(
                gameMode = GameMode.PLANECHASE,
                planechase = PlanechaseState(dieRollCount = 0),
            )
            advanceUntilIdle()

            assertEquals(0, vm.dieRollCost)
        }

        @Test
        fun `dieRollCost is 0 for first roll`() = runTest {
            val vm = makeVM()
            firestore.gameFlowSource.value = makeGame(
                gameMode = GameMode.PLANECHASE,
                planechase = PlanechaseState(dieRollCount = 1),
            )
            advanceUntilIdle()

            assertEquals(0, vm.dieRollCost)
        }

        @Test
        fun `dieRollCost increments after first`() = runTest {
            val vm = makeVM()
            firestore.gameFlowSource.value = makeGame(
                gameMode = GameMode.PLANECHASE,
                planechase = PlanechaseState(dieRollCount = 3),
            )
            advanceUntilIdle()

            assertEquals(2, vm.dieRollCost)
        }

        @Test
        fun `currentPlane returns plane from database`() = runTest {
            val plane = PlaneCard(id = "plane-1", name = "Test Plane", typeLine = "Plane", oracleText = "Test", isPhenomenon = false)
            planeDb.cards = listOf(plane)

            val vm = makeVM()
            firestore.gameFlowSource.value = makeGame(
                gameMode = GameMode.PLANECHASE,
                planechase = PlanechaseState(currentPlaneId = "plane-1"),
            )
            advanceUntilIdle()

            assertEquals(plane, vm.currentPlane)
        }

        @Test
        fun `isOwnDeckMode reads from planechase state`() = runTest {
            val vm = makeVM()
            firestore.gameFlowSource.value = makeGame(
                gameMode = GameMode.PLANECHASE,
                planechase = PlanechaseState(useOwnDeck = true),
            )
            advanceUntilIdle()

            assertTrue(vm.isOwnDeckMode)
        }

        @Test
        fun `lastDieRollerName finds player name`() = runTest {
            val vm = makeVM()
            firestore.gameFlowSource.value = makeGame(
                gameMode = GameMode.PLANECHASE,
                planechase = PlanechaseState(lastDieRollerId = "user-2"),
            )
            firestore.playersFlowSource.value = makePlayers()
            advanceUntilIdle()

            assertEquals("Player 2", vm.lastDieRollerName)
        }
    }

    // ── Identity card lookup ──

    @Nested
    inner class IdentityCardTests {

        @Test
        fun `identityCard returns card from database`() = runTest {
            val card = IdentityCard(
                id = "card-1", cardNumber = 1, name = "Test Card",
                role = "leader", abilityText = "Test", unveilCost = "3",
                rarity = "rare", hasUndercover = false,
            )
            cardDb.cards = listOf(card)

            val vm = makeVM()
            firestore.playersFlowSource.value = makePlayers()
            advanceUntilIdle()

            val player = vm.players.value.find { it.identityCardId == "card-1" }!!
            assertEquals(card, vm.identityCard(player))
        }

        @Test
        fun `identityCard returns null for player without card`() = runTest {
            val vm = makeVM()
            val player = Player(
                id = "p-x", orderId = 0, userId = "u-x", displayName = "No Card",
                lifeTotal = 40, joinedAt = now,
            )

            assertNull(vm.identityCard(player))
        }

        @Test
        fun `currentIdentityCard returns card for current player`() = runTest {
            val card = IdentityCard(
                id = "card-1", cardNumber = 1, name = "Leader Card",
                role = "leader", abilityText = "Lead", unveilCost = "3",
                rarity = "rare", hasUndercover = false,
            )
            cardDb.cards = listOf(card)

            val vm = makeVM()
            vm.currentUserId = "user-1"
            firestore.playersFlowSource.value = makePlayers()
            advanceUntilIdle()

            assertEquals(card, vm.currentIdentityCard())
        }
    }
}
